### PR TITLE
Fix Travis CI tests

### PR DESCRIPTION
**Pull request by baspeeters**
# :warning: WIP
Can't really consider it finished until https://github.com/api-platform/schema-generator/pull/124 is solved.

# Background
Build was failing on Travis CI. Every tiny fix yielded the next error.

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | no
| Fixed tickets | N/A
| License       | MIT
| Doc PR        | N/A

# Problems

## Problem 1
```
$ if [[ $lint = 1 ]]; then composer global require --dev 'phpstan/phpstan:^0.8'; fi
Changed current directory to /home/travis/.composer
./composer.json has been updated
Loading composer repositories with package information
Updating dependencies (including require-dev)
Your requirements could not be resolved to an installable set of packages.
  Problem 1
    - Conclusion: don't install phpstan/phpstan 0.8.5
...
    - Conclusion: don't install phpstan/phpstan 0.8.1
    - Conclusion: remove nikic/php-parser v4.1.0
    - Installation request for phpstan/phpstan ^0.8 -> satisfiable by phpstan/phpstan[0.8, 0.8.1, 0.8.2, 0.8.3, 0.8.4, 0.8.5].
    - Conclusion: don't install nikic/php-parser v4.1.0
    - phpstan/phpstan 0.8 requires nikic/php-parser ^3.0.2 -> satisfiable by nikic/php-parser[v3.0.2, v3.0.3, v3.0.4, v3.0.5, v3.0.6, v3.1.0, v3.1.1, v3.1.2, v3.1.3, v3.1.4, v3.1.5].
    - Can only install one of: nikic/php-parser[v3.0.2, v4.1.0].
...
    - Can only install one of: nikic/php-parser[v3.1.5, v4.1.0].
    - Installation request for nikic/php-parser (locked at v4.1.0) -> satisfiable by nikic/php-parser[v4.1.0].
```

> Installation failed, reverting ./composer.json to its original content.
> 
> The command "`if [[ $lint = 1 ]]; then composer global require --dev 'phpstan/phpstan:^0.8'; fi`" failed and exited with 2 during .
> 
> Your build has been stopped.

Fixed this by changing version `^0.8` for `phpstan/phpstan` to a higher stable one (^`0.10`).

## Problem 2
```
0.07s$ if [[ $coverage = 1 ]]; then phpdbg -qrr phpunit-6.3.phar --coverage-clover build/logs/clover.xml; else php phpunit-6.5.phar; fi
[Could not open file /home/travis/build/baspeeters/schema-generator/phpunit-6.3.phar]
[Failed to compile /home/travis/build/baspeeters/schema-generator/phpunit-6.3.phar, cannot run]
```

Fixed this by correcting the reference to PHPUnit (earlier in `.travis.yml` it explicitly installs version `6.5`).

## Problem 3
```
$ if [[ $coverage = 1 ]]; then phpdbg -qrr phpunit-6.5.phar --coverage-clover build/logs/clover.xml; else php phpunit-6.5.phar; fi
...
1) ApiPlatform\SchemaGenerator\Tests\Command\GenerateTypesCommandTest::testFluentMutators
Failed asserting that '<?php\n
\n
...
\n
    public function addFriend(Person $friend): self\n
    {\n
        $this->friends[] = $friend;\n
...
' contains "    public function addFriend(self $friend): self
    {
        $this->friends[] = $friend;
```
At first I thought this was an issue with newlines, but it appeared that this test expected the reference to a friend to be of type `self`, but this is referenced as type `Person`. PHPUnit never got to this point because of the above problems. Updated the reference to Person.

## Problem 4
Some phpstan analysis errors:
```
------ --------------------------------------- 
  Line   TypesGenerator.php                     
 ------ --------------------------------------- 
  139    Variable $graph might not be defined.  
 ------ --------------------------------------- 
```
Now the cursor gets out of the cycle when `$graph` is empty so that it won't reach the code trying to use this var which is declared in a previous foreach.
```
 ------ ------------------------------------------------------------------------- 
  Line   TypesGeneratorConfiguration.php                                          
 ------ ------------------------------------------------------------------------- 
  58     Call to an undefined method                                              
         Symfony\Component\Config\Definition\Builder\NodeDefinition::children().  
 ------ ------------------------------------------------------------------------- 
```
ArrayNode definitely has the method childred(), I've even checked it with a var dump. This took me ages. Eventually found [this issue](https://github.com/keboola/php-component/issues/8#issuecomment-370378970) where the maintainer commented also. It claims the error message is wrong and you should ignore this specific error...

## Problem 5
Installing packages with the lowest stable versions caused package `friendsofphp/php-cs-fixer` to format self-referencing entities to be formatted ad `self` instead of the entity name. Upgraded the version of this package by `0.1`.

## Problem 6
```$ composer --working-dir=build/schema-generator dump-autoload --classmap-authoritative --no-dev;

Composer could not find a composer.json file in /home/travis/build/api-platform/schema-generator/build/schema-generator
To initialize a project, please create a composer.json file as described in the https://getcomposer.org/ "Getting Started" section
Building...

  [RuntimeException]                                                                                              
  failed to get contents of '/home/travis/build/api-platform/schema-generator/build/schema-generator/bin/schema'  
```
https://github.com/api-platform/schema-generator/pull/124

### CODE REVIEW

**Comment by dunglas** [comment]
Thank you very much for handling this!

**Comment by baspeeters** [comment]
@dunglas no worries! :slightly_smiling_face: I just seem to be stuck on 1 more weird error:

```
------ ----------------------------------------------------------------------------------------------------- 
  Line   TypesGeneratorConfiguration.php                                                                      
 ------ ----------------------------------------------------------------------------------------------------- 
  65     Call to an undefined method Symfony\Component\Config\Definition\Builder\NodeDefinition::children().  
 ------ -----------------------------------------------------------------------------------------------------
```

That's what `phpstan` says in the build with `PHP: 7.2`, `coverage=1 lint=1`. It's weird because the `ArrayNode` definitely has the method `childred()`, I've even checked it with a var dump. Previously it complained about this further down the tree as well, but I was able to fix it by changing `prototype('array')` to `arrayPrototype()`, however this does not seem to work for the `$rootNode`.

**Comment by baspeeters** [comment]
Tests are green now, but there's still an error message.

**Comment by dunglas** [comment]
@baspeeters regarding the PHPStan error, it's a known issue with the Symfony Config component, you can just add it to the ignore list.

File: .travis.yml [hunk 1]
```diff
@@ -32,16 +32,19 @@ install:
...
   - |
     if [[ ! $deps && $TRAVIS_PHP_VERSION = "7.2" ]]; then
       composer install --no-dev --prefer-dist --classmap-authoritative --no-progress --no-suggest --ansi;
       ~/.composer/vendor/bin/php-scoper add-prefix --output-dir=build/schema-generator;
+      cp composer.json build/schema-generator/
+      ls -l
+      ls -l bin
       composer --working-dir=build/schema-generator dump-autoload --classmap-authoritative --no-dev;
       php -d phar.readonly=0 box.phar build;
-      php schema.phar generate-types tmp/ tests/e2e/schema.yml;
```
Comment (dunglas) [inline .travis.yml]: Can we keep it as is, it proves that the builded PHAR is working properly

File: composer.json [hunk 2]
```diff
@@ -22,13 +22,13 @@
...
         "ext-json": "*",
-        "friendsofphp/php-cs-fixer": "^2.10",
         "league/html-to-markdown": "^4.0",
         "psr/log": "^1.0",
         "symfony/config": "^3.3 || ^4.0",
         "symfony/console": "^2.7 || ^3.0 || ^4.0",
         "symfony/yaml": "^2.7 || ^3.0 || ^4.0",
-        "twig/twig": "^1.35 || ^2.0"
+        "twig/twig": "^1.35 || ^2.0",
+        "friendsofphp/php-cs-fixer": "^2.11"
```
Comment (dunglas) [inline composer.json]: Why this change? Package should be alphabetically ordered.

Comment (baspeeters) [reply composer.json]: `composer --working-dir=build/schema-generator dump-autoload --classmap-authoritative --no-dev;` was throwing errors trying to reach test files in `vendor/friendsofphp/php-cs-fixer`. I don't have full visibility as locally like the builds on Travis CI, was moving this to see what the effect was. It's already moved back.

File: .travis.yml (hunk 1 above)
Comment (baspeeters) [reply .travis.yml]: In one build it complained that it couldn't find `schema.phar`, was trying to see what happens if I changed it. Had to do trial-and-error through travis builds because I couldn't replicate it locally.

Comment (baspeeters) [reply .travis.yml]: @dunglas I've hanged this back, and as you can see it complains about

>[RuntimeException]
>failed to get contents of '/home/travis/build/api-platform/schema-generator/build/schema-generator/bin/schema'

File: composer.json (hunk 2 above)
Comment (dunglas) [reply composer.json]: @theofidry do you have an idea of what's going on here?

File: .travis.yml [hunk 3]
```diff
@@ -19,11 +19,11 @@ before_install:
   - echo "memory_limit=-1" >> ~/.phpenv/versions/$(phpenv version-name)/etc/conf.d/travis.ini
   - export PATH="$PATH:$HOME/.composer/vendor/bin"
   - wget https://phar.phpunit.de/phpunit-6.5.phar
-  - composer global require humbug/php-scoper:dev-master#674e24d2b9e90b2f629dab7f30d5070e4d49d815
```
Comment (theofidry) [inline .travis.yml]: is changing the version of PHP-Scoper necessary? It may not work that simply and if is unrelated to fixing the CI I would keep that for #124

File: composer.json (hunk 2 above)
Comment (theofidry) [reply composer.json]: Not sure but I would solve it in #124

**Comment by dunglas** [comment]
Maybe can we just drop PHP Scoper in the meantime?

**Comment by theofidry** [comment]
@dunglas done in #124 so maybe we can just ignore the PHAR related errors here

**Comment by dunglas** [comment]
Merged then, thank you very much @baspeeters
